src/group_theory/perm/cycles.lean

```diff
@@ -426,6 +436,86 @@ begin
       (ih (λ τ hτ, h1 τ (list.mem_cons_of_mem σ hτ)) (list.pairwise_of_pairwise_cons h2)) },
 end
 
+section
+variables [fintype α] {σ τ : perm α}
+
+noncomputable theory
+
+lemma is_conj_of_support_equiv (f : {x // x ∈ (σ.support : set α)} ≃ {x // x ∈ (τ.support : set α)})
+  (hf : ∀ (x : α) (hx : x ∈ (σ.support : set α)), (f ⟨σ x, apply_mem_support.2 hx⟩ : α) =
+    τ ↑(f ⟨x,hx⟩)) :
+  is_conj σ τ :=
+begin
+  have h := fintype.card_congr f,
+  rw [fintype.card_of_subtype σ.support (λ x, _), fintype.card_of_subtype τ.support (λ x, _)] at h,
+  swap, { rw finset.mem_coe },
+  swap, { rw finset.mem_coe },
+  have hc : fintype.card ((σ.support : set α)ᶜ : set α) =
+    fintype.card ((τ.support : set α)ᶜ : set α),
+  { refine (fintype.card_of_subtype σ.supportᶜ (λ x, _)).trans
+      (eq.trans _ (fintype.card_of_subtype τ.supportᶜ (λ x, _)).symm),
+    { rw [← coe_compl, finset.mem_coe] },
+    { rw [finset.card_compl, finset.card_compl, h] },
+    { rw [← coe_compl, finset.mem_coe] } },
+  rw fintype.card_eq at hc,
+  obtain ⟨cequiv⟩ := hc,
+  classical,
+  refine is_conj_iff.2 ⟨(equiv.set.sum_compl _).symm.trans ((equiv.sum_congr f cequiv).trans
+    (equiv.set.sum_compl _)), _⟩,
+  rw mul_inv_eq_iff_eq_mul,
+  ext,
+  simp only [perm.mul_apply, equiv.trans_apply, equiv.sum_congr_apply],
+  by_cases hx : x ∈ σ.support,
+  { rw [equiv.set.sum_compl_symm_apply_of_mem (finset.mem_coe.2 _), sum.map_inl,
+      equiv.set.sum_compl_apply_inl, equiv.set.sum_compl_symm_apply_of_mem (finset.mem_coe.2 _),
+      sum.map_inl, equiv.set.sum_compl_apply_inl],
+    { refine hf x (finset.mem_coe.2 _),
+      convert hx, },
+    { rw apply_mem_support,
+      convert hx } },
+  { rw [mem_support, not_not] at hx,
+    rw [equiv.set.sum_compl_symm_apply_of_not_mem, equiv.set.sum_compl_symm_apply_of_not_mem],
+    swap, { rw [finset.mem_coe, mem_support, not_not, hx] },
+    swap, { rw [finset.mem_coe, mem_support, not_not, hx, hx] },
+    simp only [sum.map_inr, set.sum_compl_apply_inr],
+    have h := (set.mem_compl_iff _ _).1 (cequiv ⟨σ x, _⟩).2,
+    rw [finset.mem_coe, mem_support, not_not, subtype.val_eq_coe, eq_comm] at h,
+    exact h.trans (congr rfl (congr rfl (congr rfl (subtype.mk_eq_mk.2 hx)))) }
+end
```

Comment:
We can make this proof a lot clearer if we broke it apart into a number of smaller definitions:

```lean
@[simp]
lemma fintype.card_subtype_compl {α : Type*} [fintype α]
  {p : α → Prop} [decidable_pred p] :
  fintype.card {x // ¬ p x} = fintype.card α - fintype.card {x // p x} :=
begin
  classical,
  rw [fintype.card_of_subtype (set.to_finset pᶜ), set.to_finset_compl p, finset.card_compl,
      fintype.card_of_subtype (set.to_finset p)];
    intros; simp; refl
end

/-- If two subtypes of a fintype have equal cardinality, so do their complements. -/
lemma fintype.card_compl_eq_card_compl {α : Type*} [fintype α] {p q : α → Prop}
  [decidable_pred p] [decidable_pred q]
  (h : fintype.card {x // p x} = fintype.card {x // q x}) :
  fintype.card {x // ¬ p x} = fintype.card {x // ¬ q x} :=
by simp only [fintype.card_subtype_compl, h]

/-- If `e` is an equivalence between two subtypes of a fintype `α`, `e.to_compl`
is an equivalence between the complement of those subtypes. -/
noncomputable def equiv.to_compl {α : Type*} [fintype α] {p q : α → Prop}
  [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) : {x // ¬ p x} ≃ {x // ¬ q x} :=
classical.choice (fintype.card_eq.mp (fintype.card_compl_eq_card_compl (fintype.card_congr e)))

/-- If `e` is an equivalence between two subtypes of a fintype `α`, `e.extend_subtype`
is a permutation of `α` acting like `e` on the subtypes and doing something arbitrary outside. -/
noncomputable def equiv.extend_subtype {α : Type*} [fintype α]
  {p q : α → Prop} [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) : perm α :=
(equiv.sum_compl p).symm.trans ((equiv.sum_congr e e.to_compl).trans
  (equiv.sum_compl q))

lemma equiv.extend_subtype_apply_of_mem {α : Type*} [fintype α]
  {p q : α → Prop} [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) (x) (hx : p x) :
  e.extend_subtype x = e ⟨x, hx⟩ :=
by { simp only [equiv.extend_subtype, equiv.trans_apply, equiv.sum_congr_apply],
     -- `p` gets turned into `λ x, p x` which `rw` doesn't like, so we have to use `erw`
     erw [equiv.set.sum_compl_symm_apply_of_mem hx, sum.map_inl,
          equiv.set.sum_compl_apply_inl q] }

lemma equiv.extend_subtype_mem {α : Type*} [fintype α]
  {p q : α → Prop} [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) (x) (hx : p x) :
  q (e.extend_subtype x) :=
by { convert (e ⟨x, hx⟩).2,
     rw [e.extend_subtype_apply_of_mem _ hx, subtype.val_eq_coe] }

lemma equiv.extend_subtype_apply_of_not_mem {α : Type*} [fintype α]
  {p q : α → Prop} [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) (x) (hx : ¬ p x) :
  e.extend_subtype x = e.to_compl ⟨x, hx⟩ :=
by { simp only [equiv.extend_subtype, equiv.trans_apply, equiv.sum_congr_apply],
    -- `p` gets turned into `λ x, p x` which `rw` doesn't like, so we have to use `erw`
    erw [equiv.set.sum_compl_symm_apply_of_not_mem hx, sum.map_inr,
         equiv.set.sum_compl_apply_inr q] }

lemma equiv.extend_subtype_not_mem {α : Type*} [fintype α]
  {p q : α → Prop} [decidable_pred p] [decidable_pred q]
  (e : {x // p x} ≃ {x // q x}) (x) (hx : ¬ p x) :
  ¬ q (e.extend_subtype x) :=
by { convert (e.to_compl ⟨x, hx⟩).2,
     rw [e.extend_subtype_apply_of_not_mem _ hx, subtype.val_eq_coe] }

lemma is_conj_of_support_equiv (f : {x // x ∈ (σ.support : set α)} ≃ {x // x ∈ (τ.support : set α)})
  (hf : ∀ (x : α) (hx : x ∈ (σ.support : set α)), (f ⟨σ x, apply_mem_support.2 hx⟩ : α) =
    τ ↑(f ⟨x,hx⟩)) :
  is_conj σ τ :=
begin
  refine is_conj_iff.2 ⟨equiv.extend_subtype f, _⟩,
  rw mul_inv_eq_iff_eq_mul,
  ext,
  simp only [perm.mul_apply],
  by_cases hx : x ∈ σ.support,
  { rw [equiv.extend_subtype_apply_of_mem, equiv.extend_subtype_apply_of_mem],
    { exact hf x (finset.mem_coe.2 hx) } },
  { have := (not_congr mem_support).mp hx,
    -- The following line I'm not so happy with, but it works:
    rw [not_not.mp this, eq_comm, ← @not_not (_ = _), ← ne.def, ← mem_support],
    exact equiv.extend_subtype_not_mem f _ hx }
end
```

Reply:
Thank you so much, this is a major improvement!